plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenFacetImpl.java

```diff
@@ -0,0 +1,426 @@
+        }
+
+        // TODO: Figure out created-by header
+        final ImmutableMap<String, String> headers = ImmutableMap.of(
+            BlobStore.BLOB_NAME_HEADER, path.getPath(),
+            BlobStore.CREATED_BY_HEADER, "unknown"
+        );
+
+        try (InputStream inputStream = content.openInputStream()) {
+          try (TempStreamSupplier supplier = new TempStreamSupplier(inputStream)) {
+            final String contentType = determineContentType(path, supplier, content.getContentType());
+            try (InputStream is = supplier.get()) {
+              tx.setBlob(is, headers, asset, HASH_ALGORITHMS, contentType);
+            }
+          }
+        }
+        tx.commit();
+      }
+      catch (IOException | InvalidContentException e) {
+        tx.rollback();
+        throw e;
+      }
+    }
+  }
+
+  @Override
+  public boolean delete(final MavenPath path) throws IOException {
+    try (StorageTx tx = getStorage().openTx()) {
+      final OrientVertex component;
+      final OrientVertex asset;
+      if (path.getCoordinates() != null) {
+        component = findComponent(tx, path.getCoordinates(), tx.getBucket());
+        if (component == null) {
+          return false;
+        }
+        asset = selectComponentAsset(tx, component, path);
+      }
+      else {
+        component = null;
+        asset = findAsset(tx, path, tx.getBucket());
+      }
+      if (asset == null) {
+        return false;
+      }
+      final boolean lastAsset = component != null && getComponentAssets(component).size() == 1;
+      tx.deleteBlob(getBlobRef(path, asset));
+      tx.deleteVertex(asset);
+      if (lastAsset) {
+        tx.deleteVertex(component);
+      }
+      tx.commit();
+      return true;
+    }
+  }
+
+  /**
+   * Returns component key based on passed in {@link Coordinates} G:A:V values.
+   */
+  private String getComponentKey(final Coordinates coordinates) {
+    // TODO: maybe sha1() the resulting string?
+    return coordinates.getGroupId()
+        + ":" + coordinates.getArtifactId()
+        + ":" + coordinates.getVersion();
+  }
+
+  /**
+   * Returns asset key based on passed in {@link MavenPath} path value.
+   */
+  private String getAssetKey(final MavenPath mavenPath) {
+    // TODO: maybe sha1() the resulting string?
+    return mavenPath.getPath();
+  }
+
+  /**
+   * Finds component by key.
+   */
+  @Nullable
+  private OrientVertex findComponent(final StorageTx tx,
+                                     final Coordinates coordinates,
+                                     final OrientVertex bucket)
+  {
+    final String componentKeyName =
+        StorageFacet.P_ATTRIBUTES + "." + getRepository().getFormat().getValue() + "." + P_COMPONENT_KEY;
+    return tx.findComponentWithProperty(componentKeyName, getComponentKey(coordinates), bucket);
+  }
+
+  /**
+   * Returns a list of all component assets.
+   */
+  private List<OrientVertex> getComponentAssets(final OrientVertex component)
```

Comment:
Agree with @mrprescott that it should use tx.findAssets(component) for now, unless my PR makes it in first :) Also, no need for the instanceof check.